src/mc-rangeopts.c

```diff
@@ -86,13 +91,96 @@ mc_RangeOpts_parse (mc_RangeOpts_t *ro,
+      {
+         if (!BSON_ITER_HOLDS_INT32 (&iter)) {
+            CLIENT_ERR ("%s'precision' must be an int32", error_prefix);
+            return false;
+         }
+         int32_t val = bson_iter_int32 (&iter);
+         if (val < 0) {
+            CLIENT_ERR ("%s'precision' must be non-negative", error_prefix);
+            return false;
+         }
+         ro->precision = OPT_U32 ((uint32_t) val);
+      }
+      END_IF_FIELD
+
       CLIENT_ERR ("%sUnrecognized field: '%s'", error_prefix, field);
       return false;
    }
 
-   CHECK_HAS (min, error_prefix);
-   CHECK_HAS (max, error_prefix);
+   // Do not error if min/max are not present. min/max are optional.
    CHECK_HAS (sparsity, error_prefix);
+   // Do not error if precision is not present. Precision is optional and only
+   // applies to double/decimal128.
+
+   // Expect precision only to be set for double or decimal128.
+   if (has_precision) {
+      if (!ro->min.set) {
+         CLIENT_ERR ("setting precision requires min");
+         return false;
+      }
+
+      bson_type_t minType = bson_iter_type (&ro->min.value);
+      if (minType != BSON_TYPE_DOUBLE && minType != BSON_TYPE_DECIMAL128) {
+         CLIENT_ERR ("expected 'precision' to be set with double or decimal128 "
+                     "index, but got: %s min",
+                     mc_bson_type_to_string (minType));
+         return false;
+      }
+
+      if (!ro->max.set) {
+         CLIENT_ERR ("setting precision requires max");
+         return false;
+      }
+
+      bson_type_t maxType = bson_iter_type (&ro->max.value);
+      if (maxType != BSON_TYPE_DOUBLE && maxType != BSON_TYPE_DECIMAL128) {
+         CLIENT_ERR ("expected 'precision' to be set with double or decimal128 "
+                     "index, but got: %s max",
+                     mc_bson_type_to_string (maxType));
+         return false;
+      }
+   }
+
+   // Expect min and max to match types.
+   if (ro->min.set && ro->max.set) {
+      bson_type_t minType = bson_iter_type (&ro->min.value),
+                  maxType = bson_iter_type (&ro->max.value);
+      if (minType != maxType) {
+         CLIENT_ERR ("expected 'min' and 'max' to be same type, but got: %s "
+                     "min and %s max",
+                     mc_bson_type_to_string (minType),
+                     mc_bson_type_to_string (maxType));
+         return false;
+      }
+   }
+
+   if (ro->min.set || ro->max.set) {
+      // Setting min/max without precision is error for double and decimal128.
+      if (ro->min.set) {
+         bson_type_t minType = bson_iter_type (&ro->min.value);
+         if (minType == BSON_TYPE_DOUBLE || minType == BSON_TYPE_DECIMAL128) {
+            if (!has_precision) {
+               CLIENT_ERR ("expected 'precision' to be set with 'min' for %s",
+                           mc_bson_type_to_string (minType));
+               return false;
+            }
+         }
+      }
+
+      if (ro->max.set) {
```

Comment:
Should there be an additional check here that `min.value <= max.value` or will that check happen in `CHECK_BOUNDS` in `mc-range-mincover.c`?

Reply:
Good question. `min.value <= max.value` is checked when encoding in `mc_getTypeInfo32`, `mc_getTypeInfo64`, and `mc_getTypeInfoDouble`.

[Test cases](https://github.com/mongodb/libmongocrypt/pull/497/commits/d908eefd62a5c9f94e0bbb0772ed2cc2872fb315) with `max > `min` were added to expect an error.